# Preprocessing IRS Tax Data for 2020

In this notebook we will be preprocessing the data for 2020 from the IRS summarizing the tax return detail by ZIP code.  Because we are conducting our analysis at a FIPS level, we will need to translate the ZIP codes into FIPS codes using a lookup.

In [1]:
import pandas as pd

In [2]:
base_df = pd.read_csv('20zpallagi.csv')

In [3]:
base_df.head()

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0          1    AL        0         1  785000.0  519980.0   85690.0  165290.0   
1          1    AL        0         2  554310.0  270870.0  121420.0  146470.0   
2          1    AL        0         3  290630.0  113280.0  124770.0   44570.0   
3          1    AL        0         4  181010.0   42010.0  120820.0   14410.0   
4          1    AL        0         5  269080.0   31310.0  224330.0    8270.0   

        ELF    CPREP  ...  N85300  A85300    N11901    A11901    N11900  \
0  724170.0  22560.0  ...     0.0     0.0   57720.0   46577.0  674840.0   
1  515150.0  13260.0  ...     0.0     0.0   81770.0  112540.0  470410.0   
2  269700.0   6420.0  ...     0.0     0.0   70360.0  144380.0  220710.0   
3  168830.0   2570.0  ...     0.0     0.0   49500.0  135429.0  130670.0   
4  252360.0   3250.0  ...   100.0    20.0  103250.0  470206.0  165650.0   

      A11900    N11902     A11902   N12000   A12000  
0  1827202.0  672200.0  1818867.0   2900.0   6089.0  
1  1445383.0  466960.0  1432458.0   4660.0  11648.0  
2   626662.0  216530.0   610170.0   5760.0  16235.0  
3   437179.0  126790.0   419324.0   3730.0  14903.0  
4   724529.0  156910.0   642895.0  11280.0  80064.0  

[5 rows x 165 columns]

In [4]:
#Drop the state detail and leave just zip code
base_df.drop(columns=['STATEFIPS', 'STATE'], inplace=True)

In [5]:
base_df.head()

zipcode  agi_stub        N1     mars1     MARS2     MARS4       ELF  \
0        0         1  785000.0  519980.0   85690.0  165290.0  724170.0   
1        0         2  554310.0  270870.0  121420.0  146470.0  515150.0   
2        0         3  290630.0  113280.0  124770.0   44570.0  269700.0   
3        0         4  181010.0   42010.0  120820.0   14410.0  168830.0   
4        0         5  269080.0   31310.0  224330.0    8270.0  252360.0   

     CPREP      PREP   DIR_DEP  ...  N85300  A85300    N11901    A11901  \
0  22560.0  394860.0  604650.0  ...     0.0     0.0   57720.0   46577.0   
1  13260.0  284210.0  431830.0  ...     0.0     0.0   81770.0  112540.0   
2   6420.0  162160.0  202590.0  ...     0.0     0.0   70360.0  144380.0   
3   2570.0  104710.0  119460.0  ...     0.0     0.0   49500.0  135429.0   
4   3250.0  157390.0  149170.0  ...   100.0    20.0  103250.0  470206.0   

     N11900     A11900    N11902     A11902   N12000   A12000  
0  674840.0  1827202.0  672200.0  1818867.0   2900.0   6089.0  
1  470410.0  1445383.0  466960.0  1432458.0   4660.0  11648.0  
2  220710.0   626662.0  216530.0   610170.0   5760.0  16235.0  
3  130670.0   437179.0  126790.0   419324.0   3730.0  14903.0  
4  165650.0   724529.0  156910.0   642895.0  11280.0  80064.0  

[5 rows x 163 columns]

In [6]:
#Import the lookup csv to translate ZIP code into FIPS code
lookup_df = pd.read_csv('FIPS_ZIP_cross_reference.csv')

In [7]:
lookup_df.head()

FIPS_CODE  ZIP_CODE
0       1001     36006
1       1001     36051
2       1001     36003
3       1001     36091
4       1001     36022

In [8]:
#Change the column order to merge
lookup_df = lookup_df[['ZIP_CODE', 'FIPS_CODE']]

In [9]:
final_df = pd.merge(left = base_df, right = lookup_df, left_on = 'zipcode', right_on = 'ZIP_CODE', how = 'left')

In [10]:
final_df.head(10)

zipcode  agi_stub        N1     mars1     MARS2     MARS4       ELF  \
0        0         1  785000.0  519980.0   85690.0  165290.0  724170.0   
1        0         2  554310.0  270870.0  121420.0  146470.0  515150.0   
2        0         3  290630.0  113280.0  124770.0   44570.0  269700.0   
3        0         4  181010.0   42010.0  120820.0   14410.0  168830.0   
4        0         5  269080.0   31310.0  224330.0    8270.0  252360.0   
5        0         6   81340.0    5000.0   72270.0     630.0   76810.0   
6    35004         1    1460.0    1010.0     200.0     210.0    1340.0   
7    35004         2    1400.0     760.0     280.0     310.0    1310.0   
8    35004         3    1020.0     460.0     350.0     190.0     970.0   
9    35004         4     630.0     140.0     420.0      60.0     590.0   

     CPREP      PREP   DIR_DEP  ...    N11901     A11901    N11900     A11900  \
0  22560.0  394860.0  604650.0  ...   57720.0    46577.0  674840.0  1827202.0   
1  13260.0  284210.0  431830.0  ...   81770.0   112540.0  470410.0  1445383.0   
2   6420.0  162160.0  202590.0  ...   70360.0   144380.0  220710.0   626662.0   
3   2570.0  104710.0  119460.0  ...   49500.0   135429.0  130670.0   437179.0   
4   3250.0  157390.0  149170.0  ...  103250.0   470206.0  165650.0   724529.0   
5    230.0   58450.0   22220.0  ...   43960.0  1216690.0   34560.0   735671.0   
6     40.0     630.0    1110.0  ...     110.0       80.0    1210.0     2807.0   
7     40.0     620.0    1120.0  ...     190.0      194.0    1210.0     3378.0   
8     30.0     460.0     750.0  ...     230.0      357.0     800.0     2131.0   
9      0.0     310.0     430.0  ...     180.0      386.0     450.0     1282.0   

     N11902     A11902   N12000    A12000  ZIP_CODE  FIPS_CODE  
0  672200.0  1818867.0   2900.0    6089.0       NaN        NaN  
1  466960.0  1432458.0   4660.0   11648.0       NaN        NaN  
2  216530.0   610170.0   5760.0   16235.0       NaN        NaN  
3  126790.0   419324.0   3730.0   14903.0       NaN        NaN  
4  156910.0   642895.0  11280.0   80064.0       NaN        NaN  
5   24770.0   285560.0   9630.0  391783.0       NaN        NaN  
6    1210.0     2795.0      0.0       0.0   35004.0     1115.0  
7    1200.0     3335.0     30.0      62.0   35004.0     1115.0  
8     790.0     2119.0      0.0       0.0   35004.0     1115.0  
9     440.0     1276.0      0.0       0.0   35004.0     1115.0  

[10 rows x 165 columns]

In [11]:
nan_df = final_df[final_df['FIPS_CODE'].isnull()]
nan_df.head(10)

zipcode  agi_stub        N1     mars1     MARS2     MARS4       ELF  \
0           0         1  785000.0  519980.0   85690.0  165290.0  724170.0   
1           0         2  554310.0  270870.0  121420.0  146470.0  515150.0   
2           0         3  290630.0  113280.0  124770.0   44570.0  269700.0   
3           0         4  181010.0   42010.0  120820.0   14410.0  168830.0   
4           0         5  269080.0   31310.0  224330.0    8270.0  252360.0   
5           0         6   81340.0    5000.0   72270.0     630.0   76810.0   
5544    99999         1    8970.0    6200.0     970.0    1690.0    7870.0   
5545    99999         2    7200.0    3990.0    1220.0    1810.0    6160.0   
5546    99999         3    3610.0    1800.0    1100.0     630.0    3280.0   
5547    99999         4    2030.0     740.0    1060.0     230.0    1870.0   

        CPREP      PREP   DIR_DEP  ...    N11901     A11901    N11900  \
0     22560.0  394860.0  604650.0  ...   57720.0    46577.0  674840.0   
1     13260.0  284210.0  431830.0  ...   81770.0   112540.0  470410.0   
2      6420.0  162160.0  202590.0  ...   70360.0   144380.0  220710.0   
3      2570.0  104710.0  119460.0  ...   49500.0   135429.0  130670.0   
4      3250.0  157390.0  149170.0  ...  103250.0   470206.0  165650.0   
5       230.0   58450.0   22220.0  ...   43960.0  1216690.0   34560.0   
5544    300.0    4960.0    5950.0  ...     680.0      530.0    7110.0   
5545    260.0    3870.0    4550.0  ...    1290.0     1805.0    5720.0   
5546    110.0    2330.0    2070.0  ...    1040.0     2364.0    2430.0   
5547     50.0    1420.0    1070.0  ...     630.0     1785.0    1230.0   

         A11900    N11902     A11902   N12000    A12000  ZIP_CODE  FIPS_CODE  
0     1827202.0  672200.0  1818867.0   2900.0    6089.0       NaN        NaN  
1     1445383.0  466960.0  1432458.0   4660.0   11648.0       NaN        NaN  
2      626662.0  216530.0   610170.0   5760.0   16235.0       NaN        NaN  
3      437179.0  126790.0   419324.0   3730.0   14903.0       NaN        NaN  
4      724529.0  156910.0   642895.0  11280.0   80064.0       NaN        NaN  
5      735671.0   24770.0   285560.0   9630.0  391783.0       NaN        NaN  
5544    16771.0    7030.0    16630.0     30.0      26.0       NaN        NaN  
5545    17087.0    5630.0    16720.0     60.0      91.0       NaN        NaN  
5546     6293.0    2330.0     5831.0     70.0     159.0       NaN        NaN  
5547     3720.0    1130.0     3256.0     60.0     160.0       NaN        NaN  

[10 rows x 165 columns]

In [12]:
nan_df['zipcode'].unique()

array([    0, 99999], dtype=int64)

Looking at the internet, the two zip codes from the data file without a FIPS code are those with zip code 0 or 99999.  Neither of these would be assigned valid zip codes anyway, from internet research they are associated with unincorporated areas that are sparsely populated.  I will drop the null rows.

In [13]:
tax_data_df = final_df[final_df['FIPS_CODE'].notnull()]

In [14]:
tax_data_df.isna().sum().sum()

0

Now that we have no null values, we can export the remaining data as a single CSV to use as a table in our database.

In [15]:
tax_data_df.to_csv('2020_tax_return_data.csv', index = False)